### PR TITLE
support nulecule with openshift

### PR DESCRIPTION
**Pull request by sub-mod**

### CODE REVIEW

**Comment by sub-mod** [comment]
depends on 
https://github.com/projectatomic/atomicapp/issues/311
https://github.com/projectatomic/atomicapp/pull/312

**Comment by sub-mod** [comment]
$ ./check-kubernetes 
mock-kube-config-basic.json                                                                                                                                                      100%  336     0.3KB/s   00:00    
.> Couldn't connect to kubernetes: Error: Couldn't find running kube-apiserver
mock-k8s-tiny-app.json                                                                                                                                                           100% 1499     1.5KB/s   00:00    
mock-k8s-tiny-app.json                                                                                                                                                           100% 1499     1.5KB/s   00:00    
mock-k8s-tiny-app.json                                                                                                                                                           100% 1499     1.5KB/s   00:00    
.> /usr/bin/atomic exited with code 127
.
----------------------------------------------------------------------
Ran 6 tests in 279.815s

OK

$ ./check-openshift
Warning: Permanently added '10.111.112.101' (ECDSA) to the list of known hosts.
config                                                                                                                                                                           100% 5653     5.5KB/s   00:00    
.Warning: Permanently added '10.111.112.101' (ECDSA) to the list of known hosts.
config                                                                                                                                                                           100% 5653     5.5KB/s   00:00    
.
----------------------------------------------------------------------
Ran 2 tests in 100.433s


testcase have passed

File: pkg/kubernetes/nulecule.js [hunk 1]
```diff
@@ -181,30 +184,96 @@ define([
...
+            var supported = false;
+            if (client.flavor === "openshift" || client.flavor === "kubernetes") {
+                supported = is_flavor_present(providersl, client.flavor);
+            } else {
+                supported = false;
+            }
+            return supported;
+        }
+
+        function versionCompare(v1, v2) {
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: This should be broken out, and unit tests should be added for it. We could eventually add it as an API to cockpit.js and add documentation, but for now putting it in kuberenetes/config.js and adding unit tests should be good enough.

File: pkg/kubernetes/nulecule.js [hunk 2]
```diff
@@ -323,6 +396,68 @@ define([
             return;
         };
 
+        function save_provider_config() {
+            var deferred = $.Deferred();
+            deferred.notify(_("saving provider config file ..."));
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: The 'S' should be capitalized here.

File: pkg/kubernetes/nulecule.js [hunk 3]
```diff
@@ -323,6 +396,68 @@ define([
             return;
         };
 
+        function save_provider_config() {
+            var deferred = $.Deferred();
+            deferred.notify(_("saving provider config file ..."));    
+            cockpit.spawn(["/usr/bin/kubectl", "config", "view", "--output=json", "--raw"])
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: Instead of launching kubectl again, I would suggest modifying the KubernetesClient so it exposes the config it connected with.

File: pkg/kubernetes/nulecule.js [hunk 4]
```diff
@@ -323,6 +396,68 @@ define([
...
+                            cfile.close();
+                            deferred.reject(new Error(_("Unable to save provider config.")));
+                        });
+                });
+            return deferred.promise();
+        }
+
+        function fetch_answers_file() {
+            var deferred = $.Deferred();
+            deferred.notify(_("reading answers file ..."));
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: First letter should be capitalized.

File: pkg/kubernetes/nulecule.js [hunk 5]
```diff
@@ -323,6 +396,68 @@ define([
...
+                        console.warn(ex);
+                    deferred.reject(new Error(_("Unable to fetch provider config.")));
+                })
+                .done(function(data) {
+                    var cfile = cockpit.file(install_dir + "/config");
+                    cfile.replace(data)
+                        .done(function (tag) {
+                            debug("provider config saved.");
+                            cfile.close();
+                            deferred.resolve();
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: Please remove extra whitespace from the ends of lines.

File: pkg/kubernetes/nulecule.js [hunk 6]
```diff
@@ -338,7 +473,8 @@ define([
                 })
                 .done(function(version_info) {
                     var atomicapp_version = version_info.atomicappversion.trim();
-                    if ( atomicapp_version == MINIMUM_SUPPORTED_ATOMICAPP_VERSION ) {
+                    var validver = versionCompare(atomicapp_version, MINIMUM_SUPPORTED_ATOMICAPP_VERSION);
+                    if ( validver > 0 ) {
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: Could you remove spaces around the condition here?

File: pkg/kubernetes/nulecule.js [hunk 7]
```diff
@@ -348,26 +484,16 @@ define([
...
+                                            .done(function(ans){
+                                                deferred.resolve(ans);
                                             })
-                                            .fail(function(ex){
+                                            .fail(function(ex) {
                                                 console.warn(ex);
                                                 deferred.reject(ex);
-                                                file.close();
+                                            })
+                                            .progress(function(msg){
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: Needs a space before parentheses.

File: test/check-openshift [hunk 8]
```diff
@@ -89,4 +89,28 @@ class TestOpenshift(MachineCase):
...
+        b.click("#deploy-app")
+        b.wait_popup("deploy-app-dialog")
+        b.set_val("#deploy-app-type", "nulecule")
+        b.set_val("#deploy-app-nulecule-image", "submod/helloapache")
+        b.set_val("#deploy-app-namespace", "mynamespace")
+        b.click("#deploy-app-start")
+        self.allow_journal_messages('Could not find any image matching "submod/helloapache".')
+        b.wait_not_attr("#deploy-app-start", "disabled", "disabled")
+        b.dialog_cancel("#deploy-app-dialog")
+
```
Comment (stefwalter) [inline test/check-openshift]: Should there be a check that the service/rc is actually deployed?

File: pkg/kubernetes/client.js [hunk 9]
```diff
@@ -497,106 +497,7 @@ define([
         };
     }
 
-    /*
-     * A helper function that returns a Promise which tries to
-     * connect to a kube-apiserver in various ways in turn.
-     */
-    function connect_api_server() {
```
Comment (stefwalter) [inline pkg/kubernetes/client.js]: Please split the code changes to client.js into a separate commit.

File: pkg/kubernetes/client.js [hunk 10]
```diff
@@ -387,7 +387,7 @@ define([
              * have to list everything again. Still watch at the same time though.
              */
             if (lastResource) {
-                params["resourceVersion"] = lastResource;
+                params.resourceVersion = lastResource;
```
Comment (stefwalter) [inline pkg/kubernetes/client.js]: This change seems unrelated to this pull request. Is there a reason it's included here?

File: pkg/kubernetes/nulecule.js [hunk 11]
```diff
@@ -323,6 +371,48 @@ define([
             return;
         };
 
+        function fetch_answers_file() {
+            var deferred = $.Deferred();
+            deferred.notify(_("Reading answers file ..."));
+            if (!client.config)
+                deferred.reject(new Error( _("Unable to read config.")));
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: This means that kubernetes successfully connected without a config file. We should fall back to kubeconfig-less mode in this case.

File: pkg/kubernetes/nulecule.js [hunk 12]
```diff
@@ -323,6 +371,48 @@ define([
...
+
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: This line has extra trailing whitespace.

File: pkg/kubernetes/client.js (hunk 10 above)
Comment (sub-mod) [inline pkg/kubernetes/client.js]: jshint was giving a warning here

**Comment by stefwalter** [comment]
I'll include these in my image rebuilds.

**Comment by stefwalter** [comment]
@sub-mod Could you rebase this? I believe images have been updated.
